compiler/symtable_.py

```diff
@@ -0,0 +1,219 @@
+        self._pos = 0   #  location of top of stack relative to ebp
+        self._items = []
+        self._sizes = {INT: 4}
+
+    def get_stack_pos(self):
+        return self._pos
+
+    def push(self, type_):
+        if type_ == INT:
+            self._items.append(INT)
+            self._pos -= self._items[INT]
+
+    def pop(self):
+        if self.is_empty():
+            raise EmptyStackException('Cannot pop empty stack!')
+        self._pos += self._sizes[self._items[-1]]
+        return self._items.pop()
+
+    def top_dtype(self):
+        if self.is_empty():
+            raise EmptyStackException('Empty Stack, nothing on top!')
+
+    def is_empty(self):
+        return bool(self._items)
+
+class OutOfScopeAccessException(Exception):
+    pass
+
+class ScopeObject:
+    def __init__(self,ScopeBlockNode,start_relative_to_prev):
+        self.ScopeBlockNode = ScopeBlockNode
+        self.start_relative_to_prev =  start_relative_to_prev
+        self.end_relative_to_start = 0
+    def incrementRelativeEnd(self,num):
+        self.end_relative_to_start += num
+
+    #Todo scope_object might not be active, take care for this problem later
+    def get_start_relative_to_scope(self,scope_object,scope_stack):
+        start_relative_to_scope_start = 0
+        count_on = False
+        multiple = 1
+        if scope_stack.is_active(scope_object) and scope_stack.is_active(self):
+            for temp in scope_stack:
+                if(temp == self):
+                    if (count_on == False):
+                        count_on = True
+                    else:
+                        count_on = False
+                if(temp == scope_object):
+                    if(count_on == False):
+                        count_on = True
+                        multiple = -1
+                    else:
+                        count_on = False
+                if(count_on == True):
+                    start_relative_to_scope_start +=  scope_stack.curren_scope().end_relative_to_start+4 #The +4 is because every scope block has ebp pushed
+        start_relative_to_scope_start *= multiple
+        return start_relative_to_scope_start
+
+GlobalScopeBlock = ScopeBlock([])
+GlobalScopeObject = ScopeObject(GlobalScopeBlock,0)
+
+class ScopeStack:
+    def __init__(self):
+        self._scopes = [GlobalScopeObject]
+
+    def enter_scope(self, scope):
+        assert (isinstance(scope, ScopeObject))
+        self._scopes.append(scope)
+
+    def exit_scope(self, scope):
+        # 'eat' scope
+        assert (scope!= GlobalScopeObject)
+        assert (self._scopes[-1]== scope)
+        self._scopes.pop()
+
+    def is_active(self, scope):
+        """Return true if 'scope' is active"""
+        assert (isinstance(scope, ScopeObject))
+        return scope in self._scopes
+
+    def current_scope(self):
+        assert (self._scopes)
+        return self._scopes[-1]
+
+
+SymbolTableEntry = namedtuple('SymbolTableEntry', ('symbol', 'type', 'scope'))
+
```

Comment:
We are using the `SymbolTableEntry` as keys for the lookup. It makes sense for them to Semantic free as possible. 
it would be better for the entry to be 
`SymbolTableEntry = namedtuple('SymbolTableEntry', ('symbol', 'scope'))`
The Symbol table should have 
`get_entry_dtype(self, entry, scope_stack)`
method which would search the scopes in order from closest to furthest for the variable and return the closest definitions `d_type` or raise an exception on failure